src/database-postgresql.cpp

```diff
@@ -0,0 +1,238 @@
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "config.h"
+
+#if USE_POSTGRESQL
+
+#include "database-postgresql.h"
+
+#include "log.h"
+#include "exceptions.h"
+#include "settings.h"
+
+/*
+ * IMPORTANT NOTICE
+ * Every conversion from non string to const char* values
+ * should be done by using an intermediate std::string
+ * buffer.
+ * If you don't use this intermediate, the conversion
+ * to std::string followed by conversion to const char*
+ * could have bad issues, refering to bad memory areas.
+ */
+
+Database_PostgreSQL::Database_PostgreSQL(const Settings &conf):
+	m_connect_string(""), m_conn(NULL), m_pgversion(0)
+{
+	if (!conf.getNoEx("pgsql_connection", m_connect_string)) {
+		throw SettingNotFoundException("Set pgsql_connection string in world.mt to "
+			"use the postgresql backend\n"
+			"Notes:\n"
+			"pgsql_connection has the following form: \n"
+			"\tpgsql_connection = host=127.0.0.1 port=5432 user=mt_user password=mt_password dbname=minetest_world\n"
+			"mt_user should have CREATE TABLE, INSERT, SELECT, UPDATE and DELETE rights on the database.\n"
+			"Don't create mt_user as a SUPERUSER!");
+	}
+
+	connectToDatabase();
+}
+
+Database_PostgreSQL::~Database_PostgreSQL()
+{
+	PQfinish(m_conn);
+}
+
+void Database_PostgreSQL::connectToDatabase()
+{
+	m_conn = PQconnectdb(m_connect_string.c_str());
+
+	if (PQstatus(m_conn) != CONNECTION_OK) {
+		throw DatabaseException(std::string(
+			"PostgreSQL database error: ") +
+			PQerrorMessage(m_conn));
+	}
+
+	m_pgversion = PQserverVersion(m_conn);
+
+	if (m_pgversion < 90500) {
+		throw DatabaseException("PostgreSQL database error: "
+			"Server version 9.5 or greater required.");
+	}
+
+	infostream << "PostgreSQL Database: Version " << m_pgversion << " Connection made." << std::endl;
+}
+
+void Database_PostgreSQL::verifyDatabase()
+{
+	if (PQstatus(m_conn) == CONNECTION_OK)
+		return;
+
+	PQreset(m_conn);
+	ping();
+}
+
+void Database_PostgreSQL::ping()
+{
+	if (PQping(m_connect_string.c_str()) != PQPING_OK) {
+		throw DatabaseException(std::string(
+			"PostgreSQL database error: ") +
+			PQerrorMessage(m_conn));
+	}
+}
+
+bool Database_PostgreSQL::initialized() const
+{
+	return (PQstatus(m_conn) == CONNECTION_OK);
+}
+
+void Database_PostgreSQL::initStatements()
+{
+	prepareStatement("read_block",
+			"SELECT block_data FROM blocks "
+					"WHERE posX = $1::int AND posY = $2::int AND posZ = $3::int");
```

Comment:
we don't need to encode this field, this is purely native and postgresql doesn't have any limitation. We don't need to compress this field data, it's just 12 bytes instead of 4.
The pg_dump on my db backup show the SQL only take 12GB of string data and 9.6 GB of raw data)
du -sh /usr/local/pgsql/data
9.6G	/usr/local/pgsql/data